### PR TITLE
nixos/htop: add module

### PR DESCRIPTION
**Pull request by mkg20001**
This module allows setting global configuration for htop in /etc/htoprc,

for example to disable showing userland threads by default

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/htop.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      defaultText = "pkgs.htop";
+      description = ''
+        The htop package that should be used.
+      '';
+    };
+
+    enable = mkEnableOption "htop process monitor";
+
+    config = mkOption {
+      type = with types; attrsOf (oneOf [ str int bool (attrsOf (oneOf [ str int bool ])) ]);
```
Comment (pennae) [inline nixos/modules/programs/htop.nix:32]: should this be
```suggestion
      type = with types; attrsOf (oneOf [ str int bool (listOf (oneOf [ str int bool ])) ]);
```

File: pkgs/tools/system/htop/default.nix [hunk 2]
```diff
@@ -32,6 +32,11 @@ stdenv.mkDerivation rec {
     ++ optional sensorsSupport "--with-sensors"
   ;
 
+  postInstall = ''
+    # link $prefix/etc to /etc (sysconfdir is $prefix/etc) to fix reading global htoprc
+    ln -s /etc $out/etc
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/system/htop/default.nix:38]: Can we adjust the path in htop instead? such symlinks are ugly.

File: nixos/modules/programs/htop.nix [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      default = pkgs.htop;
+      defaultText = "pkgs.htop";
+      description = ''
+        The htop package that should be used.
+      '';
+    };
+
+    enable = mkEnableOption "htop process monitor";
+
+    config = mkOption {
```
[the remark this answers is not in the document]
Comment (mkg20001) [reply nixos/modules/programs/htop.nix:31]: You mean putting it there?
The rewrite part is not true, as the global config is only read once as default for the userconfig

As for the parser the nix code handles that

File: pkgs/tools/system/htop/default.nix (hunk 2 above)
Comment (mkg20001) [reply pkgs/tools/system/htop/default.nix:38]: Not sure how I should set sysconfdir, seems to be `$(prefix)/etc` and prefix is basically where it should put stuff, so can't change that

Comment (SuperSandro2000) [reply pkgs/tools/system/htop/default.nix:38]: That should be referenced in code somewhere and we can hijack the path there.

File: nixos/modules/programs/htop.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/programs/htop.nix:31]: I am just quoting what the config tells you.

[the remark this answers is not in the document]
Comment (mkg20001) [reply nixos/modules/programs/htop.nix:31]: Yes. What's the meaning? It doesn't apply for global config, it doesn't seem to write there. I tested.

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/programs/htop.nix:31]: No, I mean that the config is not easy to understand and creating the right entries won't be easy.

[the remark this answers is not in the document]
Comment (mkg20001) [reply nixos/modules/programs/htop.nix:31]: It's possible to copy-and-paste things from htop userconfig after setting them in the ui. That's how I did it. Maybe could extend the example to include all the default parameters.

Comment (aanderse) [inline nixos/modules/programs/htop.nix:31]: We call this `settings` by convention.

File: pkgs/tools/system/htop/force-etc.patch [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+diff --git a/Makefile.am b/Makefile.am
+index ce50d0f..27af87e 100644
+--- a/Makefile.am
++++ b/Makefile.am
+@@ -16,7 +16,7 @@ pixmap_DATA = htop.png
+ appicondir = $(datadir)/icons/hicolor/scalable/apps
+ appicon_DATA = htop.svg
+ 
+-AM_CFLAGS += -pedantic -std=c99 -D_XOPEN_SOURCE_EXTENDED -DSYSCONFDIR="\"$(sysconfdir)\"" -I"$(top_srcdir)/$(my_htop_platform)"
```
Comment (pennae) [inline pkgs/tools/system/htop/force-etc.patch:23]: this should be accessible via the `--sysconfdir` configure flag

File: nixos/modules/programs/htop.nix [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [
+      cfg.package
+    ];
+
+    environment.etc."htoprc".text = ''
+      # Global htop configuration
+      # To change set: programs.htop.config.KEY = VALUE;
```
Comment (aanderse) [inline nixos/modules/programs/htop.nix:54]:
```suggestion
      # To change set: programs.htop.settings.KEY = VALUE;
```

**Review by aanderse — CHANGES_REQUESTED** [review]
Not really sure on the value of this module, but oh well, if it helps you I guess that is fine :man_shrugging:

**Pull request merged**